tests/test_imports.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import subprocess
+
+from accelerate.test_utils.testing import TempDirTestCase, require_tuna
+from accelerate.utils import is_tuna_interpreter_available
+
+
+if is_tuna_interpreter_available():
+    from tuna._import_profile import read_import_profile
+    from tuna_interpreter import calculate_total_time
+    from tuna_interpreter.core import get_paths_above_threshold, sort_nodes_by_total_time
+
+
+def convert_list_to_string(data):
+    end_result = ""
+    arrow_right = "->"
+    for path in data:
+        end_result += f"{arrow_right.join(path[0])} {path[1]:.3f}s\n"
+    return end_result
+
+
+def run_import_time(command: str):
+    output = subprocess.run(["python3", "-X", "importtime", "-c", command], capture_output=True, text=True)
+    return output.stderr
+
+
+@require_tuna
+class ImportSpeedTester(TempDirTestCase):
+    clear_on_setup = False
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        output = run_import_time("import torch")
+        with open(f"{cls.tmpdir}/pytorch_results.log", "w") as f:
+            f.write(output)
+        data = read_import_profile(f"{cls.tmpdir}/pytorch_results.log")
+        total_time = calculate_total_time(data)
+        cls.pytorch_time = total_time
+
+    def test_base_import(self):
+        output = run_import_time("import accelerate")
+        with open(f"{self.tmpdir}/base_results.log", "w") as f:
+            f.write(output)
+        data = read_import_profile(f"{self.tmpdir}/base_results.log")
+        total_time = calculate_total_time(data)
+        pct_more = total_time / self.pytorch_time
+        # Base import should never be more than 10% slower than raw torch import
+        err_msg = f"Base import is more than 20% slower than raw torch import ({pct_more * 100:.2f}%), please check the attached `tuna` profile:\n"
+        sorted_data = sort_nodes_by_total_time(data)
+        paths_above_threshold = get_paths_above_threshold(sorted_data, 0.1, max_depth=7)
+        err_msg += f"\n{convert_list_to_string(paths_above_threshold)}"
+        self.assertLess(pct_more, 1.2, err_msg)
```

Comment:
So IIUC, we want to ensure that the import doesn't add more than 20% on top of the PyTorch import. Did you check how robust this is? Should this test perhaps do multiple imports and then average? If so, should we throw away the first timing, as it may require a warm up (disk cache?) and thus make the first import unduly slow (and thus make the test pass trivially)?

Reply:
Correct. 

We specifically care about the warmup, because otherwise it will lead to slowdowns via `accelerate launch`. 

I'll check multiple calls, but it shouldn't because everything is run via `subprocess` so it shouldn't rely on disk caching